src/main/java/net/minestom/server/inventory/InventoryImpl.java

```diff
@@ -0,0 +1,309 @@
+     */
+    protected void handleClose(@NotNull Player player) {
+        ItemStack cursorItem = getCursorItem(player);
+
+        if (!cursorItem.isAir()) {
+            // Drop the item if it can not be added back to the inventory
+            if (!player.getInventory().addItemStack(cursorItem)) {
+                player.dropItem(cursorItem);
+            }
+        }
+
+        setCursorItem(player, ItemStack.AIR);
+        getClickPreprocessor().clearCache(player);
+        if (player.didCloseInventory()) {
+            player.UNSAFE_changeDidCloseInventory(false);
+        } else {
+            player.sendPacket(new CloseWindowPacket(getWindowId()));
+        }
+    }
+
+    /**
+     * Updates the provided slot for this inventory's viewers.
+     *
+     * @param slot the slot to update
+     * @param itemStack the item treated as in the slot
+     */
+    protected void updateSlot(int slot, @NotNull ItemStack itemStack) {
+        sendPacketToViewers(new SetSlotPacket(getWindowId(), 0, (short) slot, itemStack));
+    }
+
+    /**
+     * Updates the cursor item for the provided player.
+     *
+     * @param player the player to update
+     * @param cursorItem the cursor item to send to the player
+     */
+    protected void updateCursor(@NotNull Player player, @NotNull ItemStack cursorItem) {
+        final ItemStack currentCursorItem = cursorPlayersItem.getOrDefault(player, ItemStack.AIR);
+        if (!currentCursorItem.equals(cursorItem)) {
+            player.sendPacket(SetSlotPacket.createCursorPacket(cursorItem));
+        }
+    }
+
+    @Override
+    public void update() {
+        this.viewers.forEach(this::update);
+    }
+
+    @Override
+    public void update(@NotNull Player player) {
+        player.sendPacket(new WindowItemsPacket(getWindowId(), 0, List.of(itemStacks), cursorPlayersItem.getOrDefault(player, ItemStack.AIR)));
+    }
+
+    @Override
+    public @Nullable ClickResult handleClick(@NotNull Player player, @NotNull ClickInfo clickInfo) {
+        return DEFAULT_HANDLER.handleClick(this, player, clickInfo);
+    }
+
+    @Override
+    public @NotNull ItemStack getItemStack(int slot) {
+        return (ItemStack) ITEM_UPDATER.getVolatile(itemStacks, slot);
+    }
+
+    @Override
+    public @NotNull ItemStack[] getItemStacks() {
+        return itemStacks.clone();
+    }
+
+    @Override
+    public void copyContents(@NotNull ItemStack[] itemStacks) {
+        Check.argCondition(itemStacks.length != getSize(),
+                "The size of the array has to be of the same size as the inventory: " + getSize());
+
+        for (int i = 0; i < itemStacks.length; i++) {
+            final ItemStack itemStack = itemStacks[i];
+            Check.notNull(itemStack, "The item array cannot contain any null element!");
+            setItemStack(i, itemStack);
+        }
+    }
+
+    @Override
+    public synchronized void setItemStack(int slot, @NotNull ItemStack itemStack) {
```

Comment:
I would prefer if this PR also stopped using synchronized and instead used a lock internally.

Reply:
Will do (this is behaviour carried over from Minestom's previous inventory logic)

Reply:
I should have been more clear on what I meant here. I meant stopped using the synchronized keyword all together and instead used locks like `ReeentrantLock` (a sane default to use). 

Virtual threads do not play well with synchronized (it prevents the thread from suspending), and while Minestom hasn't been put on jdk 21 yet nor does it have any explicit support for virtual threads, I would like to move away from synchronized when related logic is being changed anyway.